### ISSUE TITLE
Unknown Error on Profile Creation (OS can't handle "special" symbols in filenames)

### ISSUE THREAD

**Issue by Snowyiu** [qTox/qTox#2275]
When creating a new profile and using the characters '|', '<' or '>' in the Username field, there'll be an unknown error when the 'Create Profile' button is clicked.

**Comment by zetok** [comment]
Well, the problem is with your OS - i.e. it can't even handle normal characters like `| < >` in filenames.
(profile name that you enter during creation of profile is used as a filename)

I guess that solution to it would be to tell the user that they can't use certain characters on error.

**Comment by ghost** [comment]
Windows' unbeaten beauty and clarity
https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx

Symbols that should be forbidden for names under Windows:
`<` `>` `:` `"` `/` `\` `|` `?` `*`

And there are more recommendations in the link.

Also important to notice is that Windows and Mac are case insensitive. In Linux at least I can create different profiles changing only the case.

@zetok on a side note: `/` is not forbidden in Linux, it can create a new user, but the profile is not saved to the disc. The same probably happens in Mac. Do you want a new issue?

**Comment by zetok** [comment]
@kehugter actually, Linux can't also handle `\0` in filenames `;)`

I think that this issue could be used to track it for all systems - I'm not sure though to what issue title could be changes to reflect that.. ideas?

**Comment by agilob** [comment]
We had the same issue in saving files sent from linux 20:12:40.jpg is not valid on Windows but is on Linux. One solution could be parse invalid filename to HTML... not pretty but will work or use boost library to do it for us... but that would increase build size and code wouldn't be nice. Remember we need to be able to save valid file and read it later, so we need both ways solution.

@antis81 ?

**Comment by ghost** [comment]
@zetok I tried it too and worked (backslash + zero)...

![qtox profile](https://cloud.githubusercontent.com/assets/6382342/10023329/d5cd55ea-6151-11e5-9c69-d3a846bc31e5.png)

![ls](https://cloud.githubusercontent.com/assets/6382342/10023331/db097750-6151-11e5-992d-bbca253c1ceb.png)

**Comment by agilob** [comment]
> @zetok I tried it too and worked (backslash + zero)...

Try the same on windows.

**Comment by zetok** [comment]
@kehugter not literal backslash + zero, but null char that it represents.

**Comment by ghost** [comment]
@zetok I see, I thought the string `\0` was meant to be interpreted as null

@agilob it will fail due to the `\` alone. And the real null too

> Use any character [...] except for the following [...] reserved characters: [...]
> Integer value zero, sometimes referred to as the ASCII NUL character.
> Characters whose integer representations are in the range from 1 through 31, except for alternate data streams where these characters are allowed. For more information about file streams, see File Streams.
> Any other character that the target file system does not allow.

**Comment by agilob** [comment]
Let's just forbid adding special characters to profile name.

**Comment by zetok** [comment]
@agilob right. Though.. I wonder about backwards compatibility - e.g. I have a profile with whitespace as a name:

``` bash
$ ls ~/.config/tox
 .ini
 .qtox_history
 .tox
03816E8F965A.ini
03816E8F965A.qtox_history
03816E8F965A.tox
```

**Comment by ghost** [comment]
The file name length restriction in Windows should also be considered for OS interportability.

**Comment by zetok** [comment]
@kehugter I wonder about that - using extended filenames (or whatever it's called) filenames on windows can be up to 32768 chars long on ntfs.

**Comment by ghost** [comment]
@zetok Up to 32767 for the full path (or almost, it can't be that easy). XP and Vista can, but not necessarily do, support it (https://msdn.microsoft.com/en-us/library/930f87yf.aspx). Those 2 OSs still make 14% of the market share according to this site
https://www.netmarketshare.com/operating-system-market-share.aspx?qprid=10&qpcustomd=0

Furthermore, most users won't be able to get to those files - http://serverfault.com/a/654756
(still an issue in Windows 10 apparently https://social.technet.microsoft.com/Forums/en-US/0e8dbb98-7bbf-4a21-bd82-7f172b867f34/path-length-still-limited-in-windows-explorer-and-other-apps?forum=WinPreview2014General)
and you can't be sure how long the available path length in Windows is - http://blog.codinghorror.com/filesystem-paths-how-long-is-too-long/

So, I guess that in Windows a better approach would be to
- restrict the path length to 256 characters
- check if the file can be written
- warn the user to chose a different name or path for the config files if it can't be

**Comment by agilob** [comment]
16 characters and restriction for special chars?

**Comment by antis81** [comment]
Excuse - wrong button…

**Comment by ghost** [comment]
Is it already decided? I see no reason to limit Linux and Mac users because of Windows. In my opinion
- `/` and `\0` should be prohibited everywhere
- In Windows, profile creation using a reserved character or too long name/path shouldn't be permitted
- In non-Windows, if the user uses a long name (>50 chars?) or Windows reserved chars, a warning should be displayed about issues with portability + explain that the profile may need to be renamed to be used in Windows
- In Linux, if the lower-case of the new profile matches a previously created one, warn that it can create portability issues in other OSs

**Comment by agilob** [comment]
> Is it already decided? I see no reason to limit Linux and Mac users because of Windows.
1. portable mode.
2. export/import profiles

**Comment by ghost** [comment]
Yes, you're right, better keep it simple (length == 16, no special chars in every OS). The nick can be changed anyway.

The current regex only permits alphanumeric chars (https://github.com/PKEv/qTox/commit/432cd4f15d4d709fa20c6c2522d5308158a5e96d#diff-1e0d837757bc75752ac84d02d7e8ed44R104). In my opinion, `-`, `_` and `.` should be permitted too. Maybe even white spaces?

@PKEv @agilob

**Comment by agilob** [comment]
I think those should be forbidden: https://en.wikipedia.org/wiki/Filename#Reserved_characters_and_words

**Comment by zetok** [comment]
> Yes, you're right, better keep it simple (length == 16, no special chars in every OS). The nick can be changed anyway.

Err, that would cause problems..

**Comment by ghost** [comment]
`-` and `_` are harmless
`.` can only be problematic in Windows if the file name ends with a dot, which doesn't happen in this case
I'm not a fan of white spaces either, but they're everywhere in Windows' paths, so they must be handled anyway

Also: should names be case insensitive, to prevent Windows errors or overwritten files?

@zetok backwards compatibility? What about qTox looking for any string as profile name but only letting the user create restricted names trough the interface?

**Comment by zetok** [comment]
@kehugter so far, backwards compatibility wasn't affected by any of PRs (at least AFAIR) - problem is though that limiting to certain chars would be a huge regression.

What would be better, is just not allowing a few certain chars to be used, without really limiting what user can use.also, as per length limit - not sure about it, I don't see it being a real problem, but then again, perhaps there would be some good, valid case to allow more than 16 chars for filename.

Small subset of chars permitted - huge regression;
Small subset of chars banned - should work well for everyone.
